### ISSUE TITLE
Extracting sprites

### ISSUE THREAD

**Issue by hanskokx** [OpenMMO/openMMO#1]
I apologize for my English, as I do not know French.  Hopefully Google Translate will get it "close enough."

Is it possible to write a BASH script to extract the sprites?  I've been trying to do it manually by following the information in the wiki, but I have not had any sort of success.  I know this project hasn't been touched in many years, but I hope you can still help me out.

---

Je m'excuse pour mon anglais, car je ne connais pas le français. Avec un peu de chance, Google Translate sera "assez proche".

Est-il possible d'écrire un script BASH pour extraire les sprites? J'ai essayé de le faire manuellement en suivant les informations du wiki, mais je n'ai eu aucun succès. Je sais que ce projet n'a pas été touché depuis de nombreuses années, mais j'espère que vous pourrez toujours m'aider.

**Comment by NewbiZ** [comment]
I don't think that's going to be easily possible. Most data files are binary and follow some ad-hoc standard to store colors and palettes.
You're better of writing a python script if C is not your thing. I have a couple lying around to extract maps.

**Comment by hanskokx** [comment]
Python is more my thing than is C.  Could you please share your examples?

**Comment by NewbiZ** [comment]
Just pushed my scripts to github/pypi: https://github.com/NewbiZ/t4c-utils

TLDR:

    $ pip install t4c-utils
    $ t4c extract -i <your install dir> -o <output dir> [-s (neerya|realmud|...)]

**Comment by hanskokx** [comment]
It was able to pull the maps out just fine, but none of the sprites could be found.

![image](https://user-images.githubusercontent.com/1911919/132998055-7f5005db-a4bf-41c8-92a1-22c72117833f.png)

Or am I reading the code correctly in that the sprites are never actually extracted?

**Comment by NewbiZ** [comment]
Sprites are not extracted yet indeed. 
I will try to add the code for their extraction soon. 

The main problem will persist though: having the sprites is not really hard, but mapping IDs to sprites is dark magic.

**Comment by NewbiZ** [comment]
Also, I cannot get how to get the proper palette to use for a sprite. It seems this mapping is voodoo magic from the client too :(

**Comment by NewbiZ** [comment]
Got motivated and pushed a version `0.0.2` that does export sprites. Some caveats though:
- Some sprites have the wrong palette. You can tweak `find_best_palette()` to choose a specific palette for a specific sprite.
- Some sprites are not properly extracted and will be deformed. I have some clues on why this happens (I think some sizes provided in the ID file need to be rounded).

I think 90% of the extracted sprites should be fine though.

    $ pip install t4c-utils==0.0.2

or

    $ pip install --upgrade t4c-utils

Tell me how that works for you

**Comment by syluna** [comment]
> Got motivated and pushed a version `0.0.2` that does export sprites. Some caveats though:
> 
> * Some sprites have the wrong palette. You can tweak `find_best_palette()` to choose a specific palette for a specific sprite.
> * Some sprites are not properly extracted and will be deformed. I have some clues on why this happens (I think some sizes provided in the ID file need to be rounded).
> 
> I think 90% of the extracted sprites should be fine though.
> 
> ```
> $ pip install t4c-utils==0.0.2
> ```
> 
> or
> 
> ```
> $ pip install --upgrade t4c-utils
> ```
> 
> Tell me how that works for you

Hello NewbiZ,

I have try your script and I have this log with error : 
```
2022-05-03 15:52:49,795 INFO Using rtmap file: C:\My Program Files\The4ThComing\neerya\game files\rt_map.dat
2022-05-03 15:52:49,817 INFO Extracting rtmap 0 to C:\My Program Files\Download\out\rtmap\world0.bmp
2022-05-03 15:52:53,318 INFO Extracting rtmap 1 to C:\My Program Files\Download\out\rtmap\world1.bmp
2022-05-03 15:52:56,780 INFO Extracting rtmap 2 to C:\My Program Files\Download\out\rtmap\world2.bmp
2022-05-03 15:53:00,260 INFO Extracting rtmap 3 to C:\My Program Files\Download\out\rtmap\world3.bmp
2022-05-03 15:53:03,735 INFO Extracting rtmap 4 to C:\My Program Files\Download\out\rtmap\world4.bmp
2022-05-03 15:53:07,211 INFO Extracting rtmap 5 to C:\My Program Files\Download\out\rtmap\world5.bmp
2022-05-03 15:53:10,657 INFO Extracting rtmap 6 to C:\My Program Files\Download\out\rtmap\world6.bmp
2022-05-03 15:53:14,144 INFO Extracting rtmap 7 to C:\My Program Files\Download\out\rtmap\world7.bmp
2022-05-03 15:53:17,608 INFO Using sprite id file: C:\My Program Files\The4ThComing\neerya\game files\v2datai.did
2022-05-03 15:53:17,609 WARNING File C:\My Program Files\The4ThComing\neerya\game files\v2datai.did does not exist
2022-05-03 15:53:17,609 WARNING Falling back to genuine C:\My Program Files\The4ThComing\game files\v2datai.did
2022-05-03 15:53:19,532 INFO Using sprite palette file: C:\My Program Files\The4ThComing\neerya\game files\v2colori.dpd
2022-05-03 15:53:19,533 WARNING File C:\My Program Files\The4ThComing\neerya\game files\v2colori.dpd does not exist
2022-05-03 15:53:19,533 WARNING Falling back to genuine C:\My Program Files\The4ThComing\game files\v2colori.dpd
Traceback (most recent call last):
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\Scripts\t4c-script.py", line 33, in <module>
    sys.exit(load_entry_point('t4c-utils==0.0.2', 'console_scripts', 't4c')())
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\t4c_utils\cli.py", line 170, in main
    cli()
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\t4c_utils\cli.py", line 54, in extract
    extract_sprites(install_directory, output_directory, server, ids, palettes)
  File "C:\Users\toto\AppData\Local\Programs\Python\Python310\lib\site-packages\t4c_utils\cli.py", line 142, in extract_sprites
    dda_idx = int(dda_idx.groupdict()['idx'])
AttributeError: 'NoneType' object has no attribute 'groupdict'
```
